tracing-fmt/src/writer.rs

```diff
@@ -0,0 +1,165 @@
+//! Abstractions for creating [`io::Write`] instances.
+
+use super::Builder;
+use std::io;
+
+/// A type that can create [`io::Write`] instances.
+///
+/// `NewWriter` is used by [`FmtSubscriber`] to print formatted text representations of [`Event`]s.
+///
+/// This trait is already implemented for function pointers and immutably-borrowing closures that
+/// return an instance of [`io::Write`], such as [`io::stdout`] and [`io::stderr`].
+///
+/// [`FmtSubscriber`]: crate::FmtSubscriber
+/// [`Event`]: tracing_core::Event
+pub trait NewWriter {
+    /// The concrete [`io::Write`] implementation returned by [`new_writer`].
+    ///
+    /// [`new_writer`]: NewWriter::new_writer
+    type Writer: io::Write;
+
+    /// Returns an instance of [`Writer`].
+    ///
+    /// # Implementer notes
+    ///
+    /// [`FmtSubscriber`] will call this method each time an event is recorded. Ensure any state
+    /// that must be saved across writes is not lost when the [`Writer`] instance is dropped. If
+    /// creating a [`io::Write`] instance is expensive, be sure to cache it when implementing
+    /// [`NewWriter`] to improve performance.
+    ///
+    /// [`Writer`]: NewWriter::Writer
+    /// [`FmtSubscriber`]: crate::FmtSubscriber
+    fn new_writer(&self) -> Self::Writer;
+}
+
+impl<F, W> NewWriter for F
+where
+    F: Fn() -> W,
+    W: io::Write,
+{
+    type Writer = W;
+
+    fn new_writer(&self) -> Self::Writer {
+        (self)()
+    }
+}
+
+impl<N, E, F, W> Builder<N, E, F, W> {
+    /// Sets the [`NewWriter`] that the subscriber being built will use to write events.
+    pub fn with_writer<W2>(self, new_writer: W2) -> Builder<N, E, F, W2>
+    where
+        W2: NewWriter + 'static,
+    {
+        Builder {
+            new_visitor: self.new_visitor,
+            fmt_event: self.fmt_event,
+            filter: self.filter,
+            settings: self.settings,
+            new_writer,
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::{FmtSubscriber, NewWriter};
+    use std::io;
+    use std::sync::{Mutex, MutexGuard, TryLockError};
+    use tracing::error;
+    use tracing_core::dispatcher::{self, Dispatch};
+
+    fn test_writer<T>(new_writer: T, msg: &str, buf: &Mutex<Vec<u8>>)
+    where
+        T: NewWriter + Send + Sync + 'static,
+    {
+        let subscriber = FmtSubscriber::builder()
+            .with_writer(new_writer)
+            .without_time()
+            .with_ansi(false)
+            .finish();
+        let dispatch = Dispatch::from(subscriber);
+
+        dispatcher::with_default(&dispatch, || {
+            error!("{}", msg);
+        });
+
+        // TODO: remove time ANSI codes when `crate::time::write` respects `with_ansi(false)`
```

Comment:
Since the presence or lack of ANSI error codes is not really what's being tested here, maybe this should just assert something like

```
let expected = format!("tracing_fmt::writer::test: {}\n", msg);
let actual = String::from_utf8(buf.try_lock().unwrap().to_vec()).unwrap();
assert_eq!(actual.contains(expected));
```

That way, the test isn't making any assertion about the presence of ANSI escapes, and will pass even when the ANSI feature flag is disabled.

I agree that the bug with `with_ansi` not disabling ANSI formatting for timestamps ought to be fixed regardless, though.